pkg/health: use configurable probing interval

jshr-w opened the pull request:
PR with core changes for #32820 / [CFP](https://github.com/cilium/design-cfps/blob/main/cilium/CFP-32820-improve-scalability-of-health-probing.md), to improve scalability of health checking.

Summary of changes:
- Update health API so healthstatusresponse contains field for probing interval.
- Make probing interval configurable via agent flag `connectivity-probe-frequency-ratio` (for preference between frequent probing vs resource use) and dynamic based on cluster size.
- Update known results/metrics at intervals of 60s, even within a longer probing interval.

```release-note
connectivity health checking: introduce dynamic probing interval based on cluster size and configurable probing frequency for improved performance at scale
```

jshr-w committed:
pkg/health: use configurable probing interval

This commit allows the user to configure a preference ratio for network
resources vs. frequency of probing in order to potentially allow for
less resource use by the prober for high-scale clusters. The ratio is a
float in [0, 1], where 0 will give most frequent probing and 1 least
frequent. This ratio is used in addition to the current cluster size to
determine a dynamic probing interval.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
api: update health model api with probing interval field

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

squeed commented on daemon/cmd/daemon_main.go:
```diff
@@ -1064,6 +1064,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.MarkHidden(option.BootIDFilename)
 	option.BindEnv(vp, option.BootIDFilename)
 
+	flags.Float64(option.ConnectivityProbeFrequencyRatio, defaults.ConnectivityProbeFrequencyRatio, "Ratio of the connectivity probe frequency vs resource usage, a float in [0, 1]. 0 will give more frequent probing, 1 will give less frequent probing. Probing frequency is dynamically adjusted based on the cluster size.")
```
(feel free to self-resolve)

We generally try and avoid adding new options to `daemon_main.go` wherever possible. However, it doesn't seem like the health "server" is a Cell yet, so you don't really have another choice.

If you felt like it, converting the health server to a Cell would be very nice indeed :-)

jshr-w replied:
Thanks for bringing this up! I would be happy to take this up in a separate PR if that seems reasonable :)

squeed commented on pkg/health/server/server.go:
```diff
@@ -364,6 +388,36 @@ func (s *Server) runActiveServices() error {
 	prober.RunLoop()
 	defer prober.Stop()
 
+	// Periodically update the cluster status, without waiting for the
+	// probing interval to pass.
```
Can you explain the reasoning behind this? (I must confess, I'm not familiar with the inner workings of the health server). If we don't have new results, what's the point in proceeding?

Is it so we correctly report "no data" for new unprobed nodes? What effect does it have on metrics?

jshr-w replied:
Yes! The idea is that the probing interval in this case could now take a longer time, especially for bigger clusters, but some results will be returned earlier than that from initial probes. So, I thought it would be nice to report the results of probes that have completed, without the user necessarily having to wait for all of them to complete.

It isn't intended to have any effect on metrics -- since there are two major metrics relevant here -- node_health_connectivity_status and node_health_connectivity_latency. The status metric is updated in every periodic update, as it is idempotent, reporting the number of endpoints with connectivity. However, the latency is a histogram metric so I needed to add a tracker of what nodes have been probed and only observe the metric if the node hasn't been seen in the same probing interval.

squeed replied:
Gotcha, thanks for the clarification.

squeed commented on pkg/health/client/testdata/allHealthy.golden:
```diff
@@ -1,2 +1,2 @@
-Cluster health:	4/4 reachable	(2023-04-01T12:00:00Z)
+Cluster health:	4/4 reachable	(2023-04-01T12:00:00Z)	(Probe interval: 1m14s)
```
Is there a point to exposing the last probe time for each node, or is it expected they will all be similar to the most recent iteration?

Would it be useful to list both start and end time for the probing run?

jshr-w replied:
I'm not sure if I understand this correctly sorry -- for exposing the probe time for each node, what is meant by 'similar to the most recent iteration'?
I kind of preferred the look of having the start time and the probe interval (instead of end time) because there may be some variation (+/- 10s, I believe) in the end time in case a probe times out. But I am open to changing it to the end time if that seems better?

squeed replied:
Perhaps my question was poorly formed.

My intuition is that there may be fairy large variation in the time a given remote host was last probed; I was wondering if it makes sense to expose this in the CLI. Something like 

```
Cluster health:	4/4 reachable	(2023-04-01T12:00:00Z)	(Probe interval: 1m14s)
Name	IP	Node	Endpoints	Last Probed
  node0	192.168.1.0	1/1	1/1	1m42s
  node1	192.168.1.1	1/1	1/1	6m0s
```

However, if you do not think this would be useful, feel free to ignore this suggestion.

jshr-w replied:
Ah, this makes sense to me! 

I've updated the PR to add this change, I've left them in separate commits to make them easier to review. I added these only to the verbose health report, as I felt it looks a bit clunky.

It looks like this now: 
```
root@kind-worker:/home/cilium# cilium-health status --verbose
Cluster health:   2/2 reachable   (2025-01-22T03:27:49Z)   (Probe interval: 2m11.83347464s)
Name              IP              Node                     Endpoints
  kind-kind/kind-worker (localhost):
    Host connectivity to 172.28.0.3:
      ICMP to stack:   OK, RTT=97.401µs    (Last probed: 2025-01-22T03:27:49Z)
      HTTP to agent:   OK, RTT=249.602µs   (Last probed: 2025-01-22T03:27:49Z)
    Secondary Host connectivity to fc00:c111::3:
      ICMP to stack:   OK, RTT=236.801µs   (Last probed: 2025-01-22T03:28:06Z)
      HTTP to agent:   OK, RTT=318.502µs   (Last probed: 2025-01-22T03:28:06Z)
    Endpoint connectivity to 10.244.1.124:
      ICMP to stack:   OK, RTT=353.702µs   (Last probed: 2025-01-22T03:28:22Z)
      HTTP to agent:   OK, RTT=457.503µs   (Last probed: 2025-01-22T03:28:22Z)
    Secondary Endpoint connectivity to fd00:10:244:1::9059:
      ICMP to stack:   OK, RTT=248.701µs   (Last probed: 2025-01-22T03:28:39Z)
      HTTP to agent:   OK, RTT=514.103µs   (Last probed: 2025-01-22T03:28:39Z)
  kind-kind/kind-control-plane:
    Host connectivity to 172.28.0.2:
      ICMP to stack:   OK, RTT=274.501µs   (Last probed: 2025-01-22T03:28:55Z)
      HTTP to agent:   OK, RTT=981.505µs   (Last probed: 2025-01-22T03:28:55Z)
    Secondary Host connectivity to fc00:c111::2:
      ICMP to stack:   OK, RTT=324.202µs   (Last probed: 2025-01-22T03:29:12Z)
      HTTP to agent:   OK, RTT=827.305µs   (Last probed: 2025-01-22T03:29:12Z)
    Endpoint connectivity to 10.244.0.166:
      ICMP to stack:   OK, RTT=448.802µs    (Last probed: 2025-01-22T03:29:28Z)
      HTTP to agent:   OK, RTT=1.222506ms   (Last probed: 2025-01-22T03:29:28Z)
    Secondary Endpoint connectivity to fd00:10:244::5945:
      ICMP to stack:   OK, RTT=417.103µs    (Last probed: 2025-01-22T03:29:45Z)
      HTTP to agent:   OK, RTT=1.229507ms   (Last probed: 2025-01-22T03:29:45Z)
```

squeed reviewed (commented):
Looks basically good, just a few questions about the details.


(My GitHub notifications are a mess, please re-request review as often as you like)

jshr-w committed:
api: update health model with per-probe last updated time

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
pkg/health: track and report last updated timestamp for probes

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

squeed commented on pkg/health/server/server.go:
```diff
@@ -364,6 +388,35 @@ func (s *Server) runActiveServices() error {
+				} else {
+					// (1) setNodes implementation doesn't override results for existing nodes.
+					// (2) Remove stale nodes so we don't report them in metrics before updating results
+					prober.setNodes(nil, nodesRemoved)
```
What happens if `nodesAdded` had a value? Would those nodes get lost?

jshr-w replied:
@squeed Good question -- I was trying to be too tricky here :) I think the cilium-health UI looks a bit messier when nodes are added in the middle of the probe interval, but I'll kick this problem to later. force pushed to update nodesAdded here as well.

squeed reviewed (approved):
Looks good. I wish I understood the node lifecycle a bit better, but that is hard to do from code review.

One question, feel free to self-resolve if not relevant.

jshr-w committed:
helm: configure health connectivity probe frequency

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

jshr-w committed:
test: set connectivityProbeFrequencyRatio to 0.0 in CI

CI uses a timeout of 4 minutes for verifying the health server status,
which is too little with the default
connectivityProbeFrequencyRatio of 0.5. Set to 0.0 to speed up initial
health checks in CI.

Signed-off-by: jshr-w <shjayaraman@microsoft.com>

Artyop reviewed (approved):
@jshr-w i've looked the issue and searched a bit on my end and I don't see any other way to manage this so lgtm
